#167377540 User can update and view profile

Nennyfills opened the pull request:
### What does this PR do?
This pull request enables a user to update and also view their profile.

#### Description of Task to be completed
- Create RESTful API route to update profile
- Create RESTful API route to see profile
- Add tests for all routes
- Add validation for routes

#### How should this be manually tested?
- git clone this branch
- run `npm install`
- Using postman, send user access token retrieved from the front end to: 
    Get  - `http://localhost:{port}/api/v1/user/profile/:id` view route
     Patch - `http://localhost:{port}/api/v1/user/profile/:id`update route

#### Any Background Context You want to provide?

#### What are the relevant pivotal tracker stories?

[#167377540](https://pivotaltracker.com/story/show/167377540)

#### Screenshots (if appropriate)
<img width="910" alt="Screenshot 2019-08-09 at 2 33 51 PM" src="https://user-images.githubusercontent.com/45515619/62782992-50dbb880-bab3-11e9-909c-1cf3ade37cd8.png">
<img width="916" alt="Screenshot 2019-08-09 at 2 33 19 PM" src="https://user-images.githubusercontent.com/45515619/62782999-5933f380-bab3-11e9-9969-9cb14a0936a3.png">
<img width="1107" alt="Screenshot 2019-08-09 at 2 29 04 PM" src="https://user-images.githubusercontent.com/45515619/62783006-5df8a780-bab3-11e9-81f3-aa76406fe23c.png">

#### Questions:

Jonathan4github reviewed (approved):
Good job

iverenshaguy commented:
> you tried

Really? I do not think this review message is professional. You tried? Please change it to something like `Good job`

iverenshaguy commented on src/controllers/user.js:
```diff
@@ -0,0 +1,59 @@
+*
+* @description allows users to update their profile
+*/
+const updateProfile = async (request) => {
```
Please export the whole controller as an object and done in the other controllers

Nennyfills replied:
Done.

iverenshaguy commented on src/controllers/user.js:
```diff
@@ -0,0 +1,59 @@
+    { ...payload },
+    { where: { id }, returning: true, raw: true }
+  );
+  if (numberOfAffectedRows === 0) return { status: 400, message: 'not updated' };
```
Good job but I'm sure the user would like to know why it wasn't updated. Also, 400 seems like a weird error status for a failed update that doesn't seem to be the fault of the user.

Nennyfills replied:
Done.

iverenshaguy commented on src/controllers/user.js:
```diff
@@ -0,0 +1,59 @@
+  const [numberOfAffectedRows, affectedRows] = await Users.update(
+    { ...payload },
+    { where: { id }, returning: true, raw: true }
+  );
```
Please leave a new line understand this block.

Nennyfills replied:
Done

iverenshaguy commented on src/controllers/user.js:
```diff
@@ -0,0 +1,59 @@
+
+  const [numberOfAffectedRows, affectedRows] = await Users.update(
+    { ...payload },
+    { where: { id }, returning: true, raw: true }
```
What would have if the user doesn't exist? Or if another user wants to update a user's profile?

Nennyfills replied:
I have added the logic for the scenario where a user doesn't exist. I will need a function from the middleware branch for the implementation of when a user updates a profile that is not theirs.

iverenshaguy commented on src/controllers/user.js:
```diff
@@ -0,0 +1,59 @@
+  const data = await Users.findOne({
+    where: { id }
+  });
+  if (!data) {
```
Leave a new line between 50 and 51 please

iverenshaguy commented on src/controllers/user.js:
```diff
@@ -0,0 +1,59 @@
+  if (!data) {
+    return { status: 404, message: 'user does not exist' };
+  }
+  delete data.dataValues.password;
```
Leave a new line between 53 and 54 please

iverenshaguy commented on src/helpers/makeUpperCase.js:
```diff
@@ -0,0 +1,18 @@
+ *
+ * @param {string} word
+ *
+ * @returns {Function} call to turn upper case to lower case
```
Good job. Please move description after the function keyword.

iverenshaguy commented on src/helpers/makeUpperCase.js:
```diff
@@ -0,0 +1,18 @@
+    .join(' ');
+  return formattedWord;
+};
+export default makeUpperCase;
```
Please leave a new line between 17 and 18

iverenshaguy commented on src/helpers/makeUpperCase.js:
```diff
@@ -0,0 +1,18 @@
+    .split(' ')
+    .map(f => f.charAt(0).toUpperCase() + f.substring(1))
+    .join(' ');
+  return formattedWord;
```
The team convention calls for a new line before `return` statements. Kindly adhere to it

iverenshaguy commented on src/helpers/responseMessage.js:
```diff
@@ -0,0 +1,17 @@
+ * @param {string} statusCode
+ * @param {string} object
+ *
+ * @returns {Void} undefined
```
```suggestion
 * @returns {void} undefined
```

iverenshaguy commented on src/helpers/responseMessage.js:
```diff
@@ -0,0 +1,17 @@
+/**
+ * Generic to response message
+ * @function
```
Leave a new line between 2 and 3

iverenshaguy commented on src/helpers/responseMessage.js:
```diff
@@ -0,0 +1,17 @@
+ * @param {string} object
+ *
+ * @returns {Void} undefined
+ * @description global response message to users
```
Move `@description` after `@function`, like a real function `@returns` should be the last thing in a JSDoc. Also, I'm not even sure our JSDocs still need to have the @description keyword since we already described it above.

iverenshaguy commented on src/helpers/responseMessage.js:
```diff
@@ -0,0 +1,17 @@
+ * @returns {Void} undefined
+ * @description global response message to users
+ */
+const responseMessage = (response, statusCode, object) => response.status(statusCode).json({
```
Good job here. However, I'm a bit confused as to why the function name is `responseMessage` and the status is `bad request`. Please look into this as I'm unable to understand why this needed to be abstracted out.

iverenshaguy commented on src/middlewares/emptybody.js:
```diff
@@ -0,0 +1,20 @@
+import responseMessage from '../helpers/responseMessage';
```
Good job. Please rename to `emptyBodyValidator` since that's what it does. the current name doesn't clearly tell me it's function.

iverenshaguy commented on src/middlewares/emptybody.js:
```diff
@@ -0,0 +1,20 @@
+ * @param {Object} response  express response object
+ * @param {Function} next  express next function
+ *
+ * @returns {Void} undefined
```
```suggestion
 * @returns {void} undefined
```

iverenshaguy commented on src/middlewares/emptybody.js:
```diff
@@ -0,0 +1,20 @@
+
+/**
+ * Generic empty body
+ * @function
```
Leave a new line before this line and remove @description

iverenshaguy commented on src/routes/index.js:
```diff
@@ -1,11 +1,13 @@
 router.use('/auth', auth);
 router.use('/articles', article);
 
+router.use('/users', user);
```
What's the use of the new line before this line?

iverenshaguy commented on src/routes/user.js:
```diff
@@ -0,0 +1,79 @@
+router.patch(
+  '/profile/:id',
+  validator(profileUpdateSchema),
+  emptybody,
```
Move the emptyBody middleware before the validator middleware. What's the need of doing the extra work when the body is empty anyways?

iverenshaguy commented on src/validations/profileSchema.js:
```diff
@@ -0,0 +1,95 @@
+      .isLength({ min: 2, max: 20 })
+      .withMessage('username must be at least 2 characters, and maximum 20')
+      .not()
+      .matches(/^[A-Za-z]+[-]{1}[A-Za-z]+([-]{1}[A-Za-z]+)+$/, 'g')
```
I'm a bit confused as to what's going on here. Also, isn't it better to use the same validation that was done for username when creating it?

iverenshaguy commented on src/validations/profileSchema.js:
```diff
@@ -0,0 +1,95 @@
+import { check } from 'express-validator/check';
+import makeUpperCase from '../helpers/makeUpperCase';
+
+const profileSchemas = {
```
Let your validations be similar to what was done for `signup`. This looks like another app. You can improve on it but they should be the same. We aren't trying to confuse the users.

Nennyfills replied:
Done.

iverenshaguy commented on src/validations/profileSchema.js:
```diff
@@ -0,0 +1,95 @@
+      .matches(/^[A-Za-z]+[-]{1}[A-Za-z]+([-]{1}[A-Za-z]+)+$/, 'g')
+      .withMessage('invalid input for username')
+      .not()
+      .matches(/^[A-Za-z]+[']+[A-Za-z]+[']+[A-Za-z]+$/, 'g')
```
Why do we have 3 different regex validations? I think it's better to use just 1 and make it as inclusive and exclusive as is needed.

Nennyfills replied:
Done.

iverenshaguy commented on src/validations/profileSchema.js:
```diff
@@ -0,0 +1,95 @@
+      .isUUID()
+      .withMessage('invalid id')
+  ],
+  profileViewSchema: [
```
Leave a new live between these sort of object properties so that the code is easy to read

Nennyfills replied:
Done

iverenshaguy replied:
For some reason, I can't see that this is done. I expect to see a new line between lines 56  and 57. Kindly recheck

iverenshaguy commented on tests/routes/profile.test.js:
```diff
@@ -0,0 +1,78 @@
+  before(async () => {
+    await Users.destroy({ cascade: true, truncate: true });
+  });
+  let newUser;
```
Declare all variables before the `before` hook. Also leave new lines around the before hook

Nennyfills replied:
Done

iverenshaguy commented on tests/routes/profile.test.js:
```diff
@@ -0,0 +1,78 @@
+  let profileData;
+
+  it('should update user', async () => {
+    profileData = {
```
Put all test data into the fixtures folder.

Nennyfills replied:
Done

iverenshaguy reviewed (changes requested):
Good job. Please review my comments

Nennyfills commented on src/validations/profileSchema.js:
```diff
@@ -0,0 +1,96 @@
+      .isLength({ min: 2, max: 20 })
+      .withMessage('firstName must be at least 2 characters, and maximum 20')
+      // .not()
+      .matches(regularExpression,)
```
Done.

iverenshaguy commented on src/helpers/makeUpperCase.js:
```diff
@@ -0,0 +1,20 @@
+ *
+ * @return {Function} call to turn upper case to lower case
+ */
+const makeUpperCase = (word) => {
```
Make uppercase? I think you mean Capitalize. This is uppercased, `UPPERCASE`, this is lowercased, `lowercase` and this is capitalized, `Capitalize`. Please correct it

Nennyfills replied:
Done.

iverenshaguy commented:
Please update all your response messages to Sentence case and make sure your tests pass on CodeClimate

Nennyfills committed:
feat(user): user can update and view profile

- create RESTful API routes to update profile
- create RESTful API routes to see profile
- add tests for all routes
- add validation for routes

[Finishes #167377540]
